src/psyclone/domain/lfric/arg_ordering.py

```diff
@@ -0,0 +1,714 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2017-2020, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors R. W. Ford, A. R. Porter and S. Siso, STFC Daresbury Lab
+# Modified I. Kavcic, Met Office
+# Modified J. Henrichs, Bureau of Meteorology
+
+'''This module implements the base class for managing arguments to
+kernel calls.
+'''
+
+from __future__ import print_function, absolute_import
+import abc
+
+from psyclone.core.access_type import AccessType
+from psyclone.domain.lfric import LFRicArgDescriptor
+from psyclone.errors import GenerationError, InternalError
+
+
+class ArgOrdering(object):
```

Comment:
I actually looked at that. The reason why I decided against is that some functions only make sense to call after `generate()` is called (e.g. `[]`, `len()`), but others must be allowed to be called before (`append`, extend()`). Considering that this would then require a fair amount of functions to be overloaded (and I somehow suspect that there are a few), I thought it wasn't worth the effort.

Am happy to redesign this, tbh I liked the idea, till I noticed the `generate()` issue.

Reply:
OK - I'm happy if you've looked into it and decided it's not worth it.